pkg/verifier/cosign/trustpolicy.go

```diff
@@ -0,0 +1,273 @@
+/*
+Copyright The Ratify Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cosign
+
+import (
+	"context"
+	"crypto"
+	"fmt"
+	"os"
+
+	re "github.com/deislabs/ratify/errors"
+	"github.com/deislabs/ratify/internal/constants"
+	"github.com/deislabs/ratify/pkg/keymanagementprovider"
+	vu "github.com/deislabs/ratify/pkg/verifier/utils"
+	"github.com/deislabs/ratify/utils"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/fulcio"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/rekor"
+	"github.com/sigstore/cosign/v2/pkg/cosign"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+)
+
+type KeyConfig struct {
+	Provider string `json:"provider,omitempty"`
+	Name     string `json:"name,omitempty"`
+	Version  string `json:"version,omitempty"`
+	File     string `json:"file,omitempty"`
+}
+
+type KeylessConfig struct {
+	RekorURL    string `json:"rekorURL,omitempty"`
+	CTLogVerify *bool  `json:"ctLogVerify,omitempty"`
+}
+
+type TrustPolicyConfig struct {
+	Name       string        `json:"name"`
+	Scopes     []string      `json:"scopes"`
+	Keys       []KeyConfig   `json:"keys,omitempty"`
+	Keyless    KeylessConfig `json:"keyless,omitempty"`
+	TLogVerify *bool         `json:"tLogVerify,omitempty"`
+}
+
+type PKKey struct {
+	Provider string `json:"provider"`
+	Name     string `json:"name,omitempty"`
+	Version  string `json:"version,omitempty"`
+}
+
+type trustPolicy struct {
+	scopes       []string
+	localKeys    map[PKKey]keymanagementprovider.PublicKey
+	config       TrustPolicyConfig
+	verifierName string
+	isKeyless    bool
+}
+
+type TrustPolicy interface {
+	GetName() string
+	GetKeys(string) (map[PKKey]keymanagementprovider.PublicKey, error)
+	GetScopes() []string
+	GetCosignOpts(context.Context) (cosign.CheckOpts, error)
+}
+
+const (
+	fileProviderName   string = "file"
+	DefaultRekorURL    string = "https://rekor.sigstore.dev"
+	DefaultTLogVerify  bool   = true
+	DefaultCTLogVerify bool   = true
+)
+
+// CreateTrustPolicy creates a trust policy from the given configuration
+// returns an error if the configuration is invalid
+// reads the public keys from the file path
+func CreateTrustPolicy(config TrustPolicyConfig, verifierName string) (TrustPolicy, error) {
+	if err := validate(config, verifierName); err != nil {
+		return nil, err
+	}
+
+	keyMap := make(map[PKKey]keymanagementprovider.PublicKey)
+	for _, keyConfig := range config.Keys {
+		// check if the key is defined by file path or by key management provider
+		if keyConfig.File != "" {
```

Comment:
curious how we handle keyMap if File == ""?

Reply:
Not sure if I'm following. Is the concern that the check is checking if it's not `""`. So potentially user could define `file` as `""` and then provider a `provider` which would be allowed by validation logic?

Reply:
ahh I see. Good question. I did consider initializing keys on cosign creation as well. The reason the logic lives in GetKeys is so that we the keyMap can be fetched everytime during verification. This is so that updates to the KMP can be picked up by the cosign verifier post verifier creation. So the flow is:

User updates KMP with new Key --> key controller updates keyMap . But cosign verifier has already been created which means trustpolicy has also been created. But we want cosign verifier to start using latest certs in keyMap which is why GetKeys will be called which will grab keys from keyMap. 

The actualy AKV or inline fetching happens on controller reconcile and not per cosign verification. Cosign verifier only interacts with keyMap. Does that answer your question?

Reply:
that makes sense to me, thanks for explanation!